src/buffer_research.py

```diff
@@ -0,0 +1,225 @@
+from src.oba_from_uniswap.common import get_block_data_file, get_prices_at_blocks
+import pickle
+
+from src.dune_api.dune_analytics import DuneAnalytics
+import pandas as pd
+from math import log
+import csv
+
+
+pd.set_option('display.max_rows', 50)
+pd.set_option("display.min_rows", 49)
+
+TOL = 1e-7
+
+total_matched_vol = 0
+total_unmatched_vol = 0
+total_rebalanced_vol = 0
+
+
+def count_number_of_saved_trades_due_to_cow(df):
+    number_of_trades_per_pair = df.groupby(
+        ["token_a_address", "token_b_address"]).size()
+    number_of_saved_trades_per_pair = dict()
+    for tf, count in number_of_trades_per_pair.items():
+        t, f = tf
+        if t < f:
+            t, f = f, t
+        if (t, f) in number_of_saved_trades_per_pair:
+            number_of_saved_trades_per_pair[t, f] += count
+        else:
+            number_of_saved_trades_per_pair[t, f] = count - 1
+    return sum(
+        number_of_saved_trades_per_pair.values())
+
+
+def apply_batch_trades_on_buffer_and_account_trade_statistic(sent_volume_per_pair, buffers, buffer_allow_listed_tokens):
```

Comment:
Yes, we could do it. But I don't expect it to change the numbers a lot. I prefer to keep it simple and trying to avoid errors

Reply:
I believe [this](https://github.com/gnosis/oba-uniswap-research/blob/master/src/oba_from_uniswap/rebalance.py#L9-L21) is the one, not sure if it will make much of a difference in the results.